app/Console/Commands/CreateMasterVoucherLogReport.php

```diff
@@ -0,0 +1,308 @@
+<?php
+
+namespace App\Console\Commands;
+
+use Carbon\Carbon;
+use DB;
+use Excel;
+use Illuminate\Support\Facades\Storage;
+use Illuminate\Console\Command;
+use Log;
+use Maatwebsite\Excel\Writers\LaravelExcelWriter;
+use ZipArchive;
+
+class CreateMasterVoucherLogReport extends Command
+{
+    /**
+     * The name and signature of the console command.
+     *
+     * @var string $signature
+     */
+    protected $signature = 'arc:createMVLReport 
+                            {--force : Execute without confirmation, eg for automation}
+                            {--no-zip : Don\'t wrap files in a single archive}
+                            {--plain : Don\'t encrypt contents of files}
+                            '
+    ;
+
+    /**
+     * The console command description.
+     *
+     * @var string $description
+     */
+    protected $description = 'Creates, encrypts and stores the MVL report under in /storage';
+
+    /**
+     * The default disk we want.
+     *
+     * @var string $disk
+     */
+    private $disk = 'enc';
+
+    /**
+     * The default archive name.
+     *
+     * @var string $archiveName
+     */
+    private $archiveName = "MVLReport.zip";
+
+    /**
+     * The sheet headers.
+     *
+     * @var array $headers
+     */
+    private $headers = [
+        'Area',
+        'Voucher Code',
+        'Dispatch Date',
+        'Disbursed Date',
+        'RVID',
+        'Participant',
+        'Payment Request Date',
+        'Trader Name',
+        'Retail Outlet',
+        'Reimbursed Date'
+    ];
+
+    /**
+     * The report's query template
+     *
+     * TODO: refactor this as eloquent lookups when it's finalised.
+     *      This will eventually help with the fact we'll need to chunk data too.
+     * @var string $report
+     */
+    private $report = <<<EOD
+SELECT
+  sponsors.name AS 'Area',
+  vouchers.code AS 'Voucher Code',
+  dispatch_date AS 'Dispatch Date',
+  disbursed_at AS 'Disbursed Date',
+  rvid AS 'RVID',
+  pri_carer_name AS 'Participant',
+  payment_request_date AS 'Payment Request Date',
+  trader_name AS 'Trader Name',
+  market_name AS 'Retail Outlet',
+  reimbursed_date AS 'Reimbursed Date'
+
+FROM vouchers
+
+  # Join for each voucher\'s sponsor name
+  LEFT JOIN sponsors ON vouchers.sponsor_id = sponsors.id
+
+  # Get our trader and market names.
+  LEFT JOIN (
+    SELECT traders.id,
+           traders.name AS trader_name,
+           markets.name AS market_name
+    FROM traders
+    LEFT JOIN markets ON traders.market_id = markets.id
+  ) AS markets_query
+    ON markets_query.id = vouchers.trader_id
+
+  # Pivot dispatched date
+  LEFT JOIN (
+    SELECT voucher_states.voucher_id,
+           voucher_states.created_at AS dispatch_date
+    FROM voucher_states
+    WHERE voucher_states.`to` = 'dispatched'
```

Comment:
Do you need the backticks?

Reply:
yes, 'to' is a reserved word in mysql and, unfortunatly the column is called that (grr, not my choice). the backticks required to be certain no strangeness occurs

Reply:
Nice.

Reply:
I'm not a big fan of that state machine. :1st_place_medal: Good wrangling though!